docs/arcadia-users-group/20231104-testing-concepts/lesson.md

```diff
@@ -0,0 +1,573 @@
+
+This lesson introduces automated testing concepts and shows how to use Python constructs to start writing tests.
+While we use Python, many of the higher-level concepts are applicable to writing tests in other languages. 
+
+<details>
+ <summary>Implementing tests in other languages</summary>
+ While this lesson uses Python, almost all programming languages have robust (and often loved) packages dedicated to testing. 
+ In R, testing is orchestrated with the <a href="https://testthat.r-lib.org/">testthat package</a>.
+ The <a href="https://r-pkgs.org">R packages</a> guide has a chapter dedicated to <a href=https://r-pkgs.org/testing-basics.html>testing</a>. 
+ Other programming languages, such as Rust, have built-in testing features.
+</details>
+
+## Lesson setup
+
+This lesson will take advantage of the skills we've learned in many previous lessons.
+We'll use [Jupyter Notebooks](https://training.arcadiascience.com/arcadia-users-group/20221024-jupyter-notebooks/lesson/), [GitHub](https://training.arcadiascience.com/workshops/20220920-intro-to-git-and-github/lesson/), [conda](https://training.arcadiascience.com/arcadia-users-group/20221017-conda/lesson/), and [Python](https://training.arcadiascience.com/arcadia-users-group/20230228-intro-to-python-1/lesson/).
+This is more overhead than we typically strive for in a lesson, but we hope that it's a chance to practice these skills to achieve a new goal.
+In the future, we may provide a [GitPod](https://www.gitpod.io/) environment for learners to use while working through this lesson, however if possible we would prefer to empower users to start implementing tests on their own computers using their own setup.
+ 
+To start this lesson, we'll begin by creating a conda environment that has the tools we'll need.
+
+```
+mamba env create -n augtest jupyter pytest 
+conda activate augtest
+```
+
+We'll also create a folder to help us stay organized.
+```
+mkdir aug-mean
+cd aug-mean
+```
+
+Once our environment is activated, start a Jupyter notebook
+
+```
+jupyter notebook
+```
+
+Now we can start learning about testing!
+
+## An introduction to testing concepts
+
+There are many ways to test software, such as assertions, exceptions, unit tests, integration tests, and regresson tests.
+
+* **Exceptions and assertions**: While writing code, `exceptions` and `assertions` can be added to sound an alarm as runtime problems come up. 
+These kinds of tests, are embedded in the software iteself and handle, as their name implies, exceptional cases rather than the norm. 
+* **Unit tests**: Unit tests investigate the behavior of units of code (such as functions, classes, or data structures).
+By validating each software unit across the valid range of its input and output parameters, tracking down unexpected behavior that may appear when the units are combined is made vastly simpler.
+* **Integration tests**: Integration tests check that various pieces of the software work together as expected. 
```

Comment:
I'd give a couple examples for the definitions of unit tests vs integration tests. 

Unit tests: a function, an individual Snakemake rule, a process in Nextflow, a cell in Jupyter
Integration tests: a set of Snakemake/Nextflow rules (or the full pipeline), the full Jupyter notebook

Reply:
nice -- added!